LibWeb: Unclosed HTML comment crashes the browser after "View Source"

SerenityOS version: commit e1598233e1c8d8326aa69871dfed85dc5d92f57c (HEAD -> master, origin/master, origin/HEAD)
(Just to confirm I am on the latest version)

(Browser = Ladybird browser)

How to trigger the crash:

1. Minimal source code: `<!--`
2. Open the file in Ladybird
3. Right click -> Select "View Source"
4. The whole browser crashes


Expected result: Show `<!--` as source


Personal conclusion: Unclosed html comment causes an error that leads to a crash of the whole browser


The key error might be:
`VERIFICATION FAILED: is_comment() at /home/me/git/serenity/Meta/Lagom/../../Userland/Libraries/LibWeb/HTML/Parser/HTMLToken.h:143`


Live demo: https://kuchikuu.xyz/ladybird/view_source_crash.html
(This page has an additional "Ok" so that the user knows the page loaded)


Log:
```
me@debianLenovo:~$ alias lb
alias lb='SERENITY_SOURCE_DIR=/home/me/git/serenity /home/me/git/serenity/Build/lagom/bin/Ladybird'
me@debianLenovo:~$ lb https://kuchikuu.xyz/ladybird/view_source_crash.html
5018.947 RequestServer(24248): Loaded 145 of 145 (100%) provided CA Certificates
5018.959 WebContent(24247): FontDatabase::load_all_fonts_from_uri('file:///home/me/.local/share/fonts'): open: No such file or directory (errno=2)
5021.455 WebContent(24247): ResourceLoader: Failed load of: "https://kuchikuu.xyz/favicon.ico", Error: Load failed: 404, Duration: 392ms
VERIFICATION FAILED: is_comment() at /home/me/git/serenity/Meta/Lagom/../../Userland/Libraries/LibWeb/HTML/Parser/HTMLToken.h:143
/home/me/git/serenity/Build/lagom/lib/liblagom-ak.so.0(ak_verification_failed+0xad) [0x7f0b0ef7bf0d]
/home/me/git/serenity/Build/lagom/lib/liblagom-web.so.0 Web::HTML::HTMLTokenizer::next_token(Web::HTML::HTMLTokenizer::StopAtInsertionPoint) 0xeb50) [0x7f0b0c6b0b20]
/home/me/git/serenity/Build/lagom/lib/liblagom-webview.so.0 WebView::highlight_source(URL::URL const&, AK::StringView) 0x2d0) [0x7f0b0f341fb0]
/home/me/git/serenity/Build/lagom/bin/Ladybird(+0x4011a) [0x55b6408eb11a]
/home/me/git/serenity/Build/lagom/lib/liblagom-webview.so.0 WebView::WebContentClient::did_get_source(unsigned long, URL::URL const&, AK::ByteString const&) 0xec) [0x7f0b0f35c77c]
/home/me/git/serenity/Build/lagom/lib/liblagom-webview.so.0(+0xef944) [0x7f0b0f399944]
/home/me/git/serenity/Build/lagom/lib/liblagom-ipc.so.0 IPC::ConnectionBase::handle_messages() 0xb2) [0x7f0b0f06f722]
/home/me/git/serenity/Build/lagom/lib/liblagom-webview.so.0(+0xbb910) [0x7f0b0f365910]
/home/me/git/serenity/Build/lagom/lib/liblagom-core.so.0(+0x4dd34) [0x7f0b0f035d34]
/home/me/git/serenity/Build/lagom/lib/liblagom-core.so.0 Core::Notifier::event(Core::Event&) 0x62) [0x7f0b0f02a6f2]
/home/me/git/serenity/Build/lagom/lib/liblagom-core.so.0 Core::EventReceiver::dispatch_event(Core::Event&, Core::EventReceiver*) 0x70) [0x7f0b0f0251f0]
/home/me/git/serenity/Build/lagom/bin/Ladybird(+0x38878) [0x55b6408e3878]
/lib/x86_64-linux-gnu/libQt6Core.so.6(+0x16cb36) [0x7f0b0db6cb36]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QSocketNotifier::activated(QSocketDescriptor, QSocketNotifier::Type, QSocketNotifier::QPrivateSignal) 0x5f) [0x7f0b0db70bdf]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QSocketNotifier::event(QEvent*) 0x8b) [0x7f0b0db7316b]
/lib/x86_64-linux-gnu/libQt6Widgets.so.6 QApplicationPrivate::notify_helper(QObject*, QEvent*) 0x93) [0x7f0b0e982a53]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QCoreApplication::notifyInternal2(QObject*, QEvent*) 0x118) [0x7f0b0db238b8]
/lib/x86_64-linux-gnu/libQt6Core.so.6(+0x31033f) [0x7f0b0dd1033f]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_context_dispatch+0x299) [0x7f0b0a7da7a9]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(+0x54a38) [0x7f0b0a7daa38]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_context_iteration+0x2c) [0x7f0b0a7daacc]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) 0x6c) [0x7f0b0dd07f7c]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) 0x2aa) [0x7f0b0db2d57a]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QCoreApplication::exec() 0x98) [0x7f0b0db266a8]
/home/me/git/serenity/Build/lagom/lib/liblagom-core.so.0 Core::EventLoop::exec() 0x36) [0x7f0b0f01e9d6]
/home/me/git/serenity/Build/lagom/bin/Ladybird(+0x57f9f) [0x55b640902f9f]
/home/me/git/serenity/Build/lagom/bin/Ladybird(+0x21a28) [0x55b6408cca28]
/lib/x86_64-linux-gnu/libc.so.6(+0x2724a) [0x7f0b0b84624a]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x85) [0x7f0b0b846305]
/home/me/git/serenity/Build/lagom/bin/Ladybird(+0x21b41) [0x55b6408ccb41]
Illegal instruction
```

And here is the log from ./Meta/Serenity.sh just in case anyone wants it

```
me@debianLenovo:~/git/serenity$ ./Meta/serenity.sh run lagom ladybird https://kuchikuu.xyz/ladybird/view_source_crash.html
-- Configuring done
-- Generating done
-- Build files have been written to: /home/me/git/serenity/Build/lagom
ninja: Entering directory `/home/me/git/serenity/Build/lagom'
[0/2] Re-checking globbed directories...
[0/1] cd /home/me/git/serenity/Build/lagom/Ladybird && /usr/bin/cmake -E env SERENITY_SOURCE_...me/git/serenity/Build/lagom/bin/Ladybird https://kuchikuu.xyz/ladybird/view_source_crash.html
5291.711 RequestServer(24422): Loaded 145 of 145 (100%) provided CA Certificates
5291.727 WebContent(24421): FontDatabase::load_all_fonts_from_uri('file:///home/me/.local/share/fonts'): open: No such file or directory (errno=2)
5294.490 WebContent(24421): ResourceLoader: Failed load of: "https://kuchikuu.xyz/favicon.ico", Error: Load failed: 404, Duration: 420ms
VERIFICATION FAILED: is_comment() at /home/me/git/serenity/Meta/Lagom/../../Userland/Libraries/LibWeb/HTML/Parser/HTMLToken.h:143
/home/me/git/serenity/Build/lagom/lib/liblagom-ak.so.0(ak_verification_failed+0xad) [0x7fced2ba9f0d]
/home/me/git/serenity/Build/lagom/lib/liblagom-web.so.0 Web::HTML::HTMLTokenizer::next_token(Web::HTML::HTMLTokenizer::StopAtInsertionPoint) 0xeb50) [0x7fced0ab0b20]
/home/me/git/serenity/Build/lagom/lib/liblagom-webview.so.0 WebView::highlight_source(URL::URL const&, AK::StringView) 0x2d0) [0x7fced36d8fb0]
/home/me/git/serenity/Build/lagom/bin/Ladybird(+0x4011a) [0x564fe51ba11a]
/home/me/git/serenity/Build/lagom/lib/liblagom-webview.so.0 WebView::WebContentClient::did_get_source(unsigned long, URL::URL const&, AK::ByteString const&) 0xec) [0x7fced36f377c]
/home/me/git/serenity/Build/lagom/lib/liblagom-webview.so.0(+0xef944) [0x7fced3730944]
/home/me/git/serenity/Build/lagom/lib/liblagom-ipc.so.0 IPC::ConnectionBase::handle_messages() 0xb2) [0x7fced3406722]
/home/me/git/serenity/Build/lagom/lib/liblagom-webview.so.0(+0xbb910) [0x7fced36fc910]
/home/me/git/serenity/Build/lagom/lib/liblagom-core.so.0(+0x4dd34) [0x7fced33ccd34]
/home/me/git/serenity/Build/lagom/lib/liblagom-core.so.0 Core::Notifier::event(Core::Event&) 0x62) [0x7fced33c16f2]
/home/me/git/serenity/Build/lagom/lib/liblagom-core.so.0 Core::EventReceiver::dispatch_event(Core::Event&, Core::EventReceiver*) 0x70) [0x7fced33bc1f0]
/home/me/git/serenity/Build/lagom/bin/Ladybird(+0x38878) [0x564fe51b2878]
/lib/x86_64-linux-gnu/libQt6Core.so.6(+0x16cb36) [0x7fced1f6cb36]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QSocketNotifier::activated(QSocketDescriptor, QSocketNotifier::Type, QSocketNotifier::QPrivateSignal) 0x5f) [0x7fced1f70bdf]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QSocketNotifier::event(QEvent*) 0x8b) [0x7fced1f7316b]
/lib/x86_64-linux-gnu/libQt6Widgets.so.6 QApplicationPrivate::notify_helper(QObject*, QEvent*) 0x93) [0x7fced2d82a53]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QCoreApplication::notifyInternal2(QObject*, QEvent*) 0x118) [0x7fced1f238b8]
/lib/x86_64-linux-gnu/libQt6Core.so.6(+0x31033f) [0x7fced211033f]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_context_dispatch+0x299) [0x7fcecebda7a9]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(+0x54a38) [0x7fcecebdaa38]
/lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_context_iteration+0x2c) [0x7fcecebdaacc]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) 0x6c) [0x7fced2107f7c]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) 0x2aa) [0x7fced1f2d57a]
/lib/x86_64-linux-gnu/libQt6Core.so.6 QCoreApplication::exec() 0x98) [0x7fced1f266a8]
/home/me/git/serenity/Build/lagom/lib/liblagom-core.so.0 Core::EventLoop::exec() 0x36) [0x7fced33b59d6]
/home/me/git/serenity/Build/lagom/bin/Ladybird(+0x57f9f) [0x564fe51d1f9f]
/home/me/git/serenity/Build/lagom/bin/Ladybird(+0x21a28) [0x564fe519ba28]
/lib/x86_64-linux-gnu/libc.so.6(+0x2724a) [0x7fcecfc4624a]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x85) [0x7fcecfc46305]
/home/me/git/serenity/Build/lagom/bin/Ladybird(+0x21b41) [0x564fe519bb41]
Illegal instruction
FAILED: Ladybird/CMakeFiles/run-ladybird /home/me/git/serenity/Build/lagom/Ladybird/CMakeFiles/run-ladybird 
cd /home/me/git/serenity/Build/lagom/Ladybird && /usr/bin/cmake -E env SERENITY_SOURCE_DIR=/home/me/git/serenity /home/me/git/serenity/Build/lagom/bin/Ladybird https://kuchikuu.xyz/ladybird/view_source_crash.html
ninja: build stopped: subcommand failed.
```

also happens in commit e232a84f0ef1e19accf65dc7a7ac5e4d9223a183 (randomly selected older version for a quick test)

--- a/Userland/Libraries/LibWeb/HTML/Parser/HTMLTokenizer.cpp
+++ b/Userland/Libraries/LibWeb/HTML/Parser/HTMLTokenizer.cpp
@@ -137,11 +137,17 @@ namespace Web::HTML {
         return m_queued_tokens.dequeue();               \
     } while (0)
 
-#define EMIT_CURRENT_TOKEN                              \
+#define EMIT_CURRENT_TOKEN_FOLLOWED_BY_EOF              \
     do {                                                \
         VERIFY(m_current_builder.is_empty());           \
         will_emit(m_current_token);                     \
         m_queued_tokens.enqueue(move(m_current_token)); \
+                                                        \
+        m_has_emitted_eof = true;                       \
+        create_new_token(HTMLToken::Type::EndOfFile);   \
+        will_emit(m_current_token);                     \
+        m_queued_tokens.enqueue(move(m_current_token)); \
+                                                        \
         return m_queued_tokens.dequeue();               \
     } while (0)
 
@@ -1428,7 +1434,7 @@ _StartOfFunction:
                 ON_EOF
                 {
                     log_parse_error();
-                    EMIT_EOF;
+                    EMIT_CURRENT_TOKEN_FOLLOWED_BY_EOF;
                 }
                 ANYTHING_ELSE
                 {
@@ -1460,7 +1466,7 @@ _StartOfFunction:
                 {
                     log_parse_error();
                     m_current_token.set_comment(consume_current_builder());
-                    EMIT_EOF;
+                    EMIT_CURRENT_TOKEN_FOLLOWED_BY_EOF;
                 }
                 ANYTHING_ELSE
                 {
@@ -1491,7 +1497,7 @@ _StartOfFunction:
                 {
                     log_parse_error();
                     m_current_token.set_comment(consume_current_builder());
-                    EMIT_EOF;
+                    EMIT_CURRENT_TOKEN_FOLLOWED_BY_EOF;
                 }
                 ANYTHING_ELSE
                 {
@@ -1519,7 +1525,7 @@ _StartOfFunction:
                 {
                     log_parse_error();
                     m_current_token.set_comment(consume_current_builder());
-                    EMIT_EOF;
+                    EMIT_CURRENT_TOKEN_FOLLOWED_BY_EOF;
                 }
                 ANYTHING_ELSE
                 {
@@ -1540,7 +1546,7 @@ _StartOfFunction:
                 {
                     log_parse_error();
                     m_current_token.set_comment(consume_current_builder());
-                    EMIT_EOF;
+                    EMIT_CURRENT_TOKEN_FOLLOWED_BY_EOF;
                 }
                 ANYTHING_ELSE
                 {

--- a/Userland/Libraries/LibWebView/SourceHighlighter.cpp
+++ b/Userland/Libraries/LibWebView/SourceHighlighter.cpp
@@ -60,7 +60,7 @@ String highlight_source(URL::URL const& url, StringView source)
         previous_position = end_position;
     };
 
-    for (auto token = tokenizer.next_token(); token.has_value(); token = tokenizer.next_token()) {
+    for (auto token = tokenizer.next_token(); token.has_value() && !token->is_end_of_file(); token = tokenizer.next_token()) {
         if (token->is_comment()) {
             append_source(token->start_position().byte_offset);
             append_source(token->end_position().byte_offset, "comment"sv);

